lib/constructs/identity-provider.ts

```diff
@@ -0,0 +1,250 @@
+  domainPrefix?: string;
+}
+
+/**
+ * An internal identity provider for the application.
+ */
+export class CognitoIdentityProvider extends Construct implements IIdentityProvider {
+  public readonly domainName: string;
+
+  public readonly userPool: cognito.IUserPool;
+  public readonly domain: cognito.IUserPoolDomain;
+
+  private readonly resourceServers: cognito.UserPoolResourceServer[] = [];
+  constructor(scope: Construct, id: string, props?: CognitoIdentityProviderProps) {
+    super(scope, id);
+    this.userPool = new cognito.UserPool(this, "IdP", {});
+    this.domain = this.userPool.addDomain("Domain", {
+      cognitoDomain: {
+        domainPrefix: props?.domainPrefix ?? id.toLowerCase(),
+      },
+    });
+    this.domainName = this.domain.domainName;
+    if (props?.scopes) {
+      for (const scopeConfig of props.scopes) {
+        if (!scopeConfig.resourceServerName) throw new Error("Cognito requires a resource server");
+        this.resourceServers.push(
+          this.userPool.addResourceServer(scopeConfig.resourceServerName, {
+            identifier: scopeConfig.resourceServerName,
+            scopes: scopeConfig.scopes.map(
+              (scope) => new cognito.ResourceServerScope({ scopeName: scope.name, scopeDescription: scope.description })
+            ),
+          })
+        );
+      }
+    }
+  }
+
+  public addClient(app: IIdentityProviderClientApplication, scopes: string[]): IIdentityProviderClient {
+    const client = new CognitoIdentityProviderClient(this, `${app.name}Client`, {
+      userPool: this.userPool,
+      clientApplication: app,
+      scopes,
+    });
+    // We can pretty safely perform this because custom domains are not support in GovCloud
+    // so we can be pretty sure that we don't have a full custom domain here, just a prefix.
+    // Ideally, we'd have some way to detect that and provide a slightly better method to
+    // build this URL than knowing about the internals of how Cogntio URLs are generated,
+    // honestly ideally that information would just come from the CDK itself.
+    // TODO: Make this work in multiple partitions (standard, ISO) and regions or with
+    // custom domains.
+    const fullDomain = `${this.domain.domainName}.auth-fips.${cdk.Aws.REGION}.amazoncognito.com`;
+    app.configure(client.clientId, client.secret, fullDomain);
+    return client;
+  }
+}
+
+interface UserPoolClientDetailsProps {
+  client: cognito.IUserPoolClient;
+  userPool: cognito.IUserPool;
+}
+
+/**
+ * A custom resource with additional details about an AWS::Cognito::UserPoolClient.
+ *
+ * This AWS::Cognito::UserPoolClient resource type exposes depressingly few attributes
+ * about the Client itself (like, not even the ID). This custom resource performs
+ * a cognito-idp:DescribeUserPoolClient API call to fetch all the attributes. This
+ * may not be a totally-ideal use of a custom resource but it means that we don't have
```

Comment:
I challenge this hyphenation!
```suggestion
 * may not be a totally ideal use of a custom resource but it means that we don't have
```